pkg/cli/commands.go

```diff
@@ -44,12 +61,77 @@ func New() *cobra.Command {
 				slog.SetDefault(slog.New(charmlog.NewWithOptions(os.Stderr, charmlog.Options{ReportTimestamp: true, Level: charmlog.Level(level)})))
 			}
 
+			var (
+				err error
+				exp sdktrace.SpanExporter
+			)
+
+			if gcptrace {
+				exp, err = gtrace.New()
+				if err != nil {
+					return fmt.Errorf("creating gcp trace exporter: %w", err)
+				}
+			} else if traceFile != "" {
+				w, err := os.Create(traceFile)
+				if err != nil {
+					return fmt.Errorf("creating trace file: %w", err)
+				}
+
+				deferred(func(context.Context) error {
+					return w.Close()
+				})
+
+				exp, err = stdouttrace.New(stdouttrace.WithWriter(w))
+				if err != nil {
+					return fmt.Errorf("creating stdout exporter: %w", err)
+				}
+			}
+
+			if exp != nil {
+				tp := sdktrace.NewTracerProvider(sdktrace.WithBatcher(exp))
+				otel.SetTracerProvider(tp)
+
+				deferred(func(ctx context.Context) error {
+					return tp.Shutdown(ctx)
+				})
+
+				deferred(func(ctx context.Context) error {
+					return tp.ForceFlush(ctx)
+				})
+
+				ctx, span := otel.Tracer("melange").Start(cmd.Context(), cmd.CalledAs())
+				cmd.SetContext(ctx)
+
+				deferred(func(context.Context) error {
+					span.End()
+					return nil
+				})
+			}
+
 			return nil
 		},
+		PersistentPostRunE: func(cmd *cobra.Command, args []string) error {
+			// It shouldn't take very long to flush spans but we don't want the whole build to fail if it takes a few seconds.
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			defer cancel()
+
+			slices.Reverse(stack)
+
+			errs := []error{}
+			for _, f := range stack {
+				errs = append(errs, f(ctx))
+			}
+
+			return errors.Join(errs...)
+		},
 	}
 	cmd.PersistentFlags().Var(&level, "log-level", "log level (e.g. debug, info, warn, error)")
 	cmd.PersistentFlags().BoolVar(&gcplog, "gcplog", false, "use GCP logging")
 	_ = cmd.PersistentFlags().MarkHidden("gcplog")
+	cmd.PersistentFlags().BoolVar(&gcptrace, "gcptrace", false, "use GCP tracing")
+	_ = cmd.PersistentFlags().MarkHidden("gcptrace")
+	cmd.PersistentFlags().StringVar(&traceFile, "trace", "", "where to write trace output")
+	_ = cmd.PersistentFlags().MarkHidden("trace")
```

Comment:
If we're marking this as hidden, could we have some documentation somewhere to show that it exists and how to use it? It's an incredible developer feature, IMHO

Reply:
I think basically nobody but me uses it, and I'd like to be able to break it in the future.

Reply:
I'd like to explicitly not commit to supporting this right now, which means (unfortunately) not telling anybody outside of this completely private circle of trust that it exists.

If it proves as useful as we think it is, and word catches on, we should un-hide it and document it.

I think a more ideal outcome would be that all the GCP trace exporting happens in a different, non-melange-CLI caller that wraps melange-as-a-library, since running the melange CLI in GCP directly as a long-term strategy is Bonkers™️.

Reply:
I agree with that more ideal outcome. And I agree that we shouldn't overcommit to supporting something externally.

I'm okay with leaving it hidden. I'd like to make a non-blocking request that we at least document this internally, if not now then soon. I'd love to move away from things that only one person gets to know about, even when the thing is not yet considered "stable".